nav2_costmap_2d/test/integration/test_costmap_2d_publisher.cpp

```diff
@@ -0,0 +1,151 @@
+      rclcpp::CallbackGroupType::MutuallyExclusive, false);
+
+    rclcpp::SubscriptionOptions sub_option;
+    sub_option.callback_group = callback_group_;
+
+    std::string topic_name = "/dummy_costmap/static_layer_raw";
+    costmap_sub_ = this->create_subscription<nav2_msgs::msg::Costmap>(
+      topic_name,
+      rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+      std::bind(&TestCostmap2dPublisher::costmapCallback, this, std::placeholders::_1),
+      sub_option);
+
+    executor_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+    executor_->add_callback_group(callback_group_, get_node_base_interface());
+    executor_thread_ = std::make_unique<nav2_util::NodeThread>(executor_);
+
+    costmap_ros_ = std::make_shared<nav2_costmap_2d::Costmap2DROS>(
+      "dummy_costmap",
+      std::string{get_namespace()},
+      "dummy_costmap",
+      get_parameter("use_sim_time").as_bool());
+    costmap_thread_ = std::make_unique<nav2_util::NodeThread>(costmap_ros_);
+
+    costmap_ros_->configure();
+
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  on_activate(const rclcpp_lifecycle::State &) override
+  {
+    RCLCPP_INFO(get_logger(), "Activating");
+    costmap_ros_->activate();
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  on_deactivate(const rclcpp_lifecycle::State &) override
+  {
+    RCLCPP_INFO(get_logger(), "Deactivating");
+    costmap_ros_->deactivate();
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  on_cleanup(const rclcpp_lifecycle::State &) override
+  {
+    executor_thread_.reset();
+    costmap_thread_.reset();
+    costmap_ros_->deactivate();
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  void costmapCallback(const nav2_msgs::msg::Costmap::SharedPtr costmap)
+  {
+    promise_.set_value(costmap);
+  }
+
+  rclcpp::Subscription<nav2_msgs::msg::Costmap>::SharedPtr costmap_sub_;
+
+  rclcpp::CallbackGroup::SharedPtr callback_group_;
+  rclcpp::executors::SingleThreadedExecutor::SharedPtr executor_;
+  std::unique_ptr<nav2_util::NodeThread> executor_thread_;
+  std::promise<nav2_msgs::msg::Costmap::SharedPtr> promise_;
+
+  std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros_;
+  std::unique_ptr<nav2_util::NodeThread> costmap_thread_;
+};
+
+class TestNode : public ::testing::Test
+{
+public:
+  TestNode()
+  {
+    costmap_publisher_ = std::make_shared<TestCostmap2dPublisher>("test_costmap_publisher");
+    costmap_publisher_->on_configure(costmap_publisher_->get_current_state());
+    costmap_publisher_->on_activate(costmap_publisher_->get_current_state());
+  }
+
+  ~TestNode() override
+  {
+    costmap_publisher_->on_deactivate(costmap_publisher_->get_current_state());
+    costmap_publisher_->on_cleanup(costmap_publisher_->get_current_state());
+  }
+
+protected:
+  std::shared_ptr<TestCostmap2dPublisher> costmap_publisher_;
```

Comment:
Move the costmap 2d ros object into the TestNode. Its quite confusing to have it in the publisher fixture. I think that separates your intent much more cleanly.